keps/draft-20180426-namespace-template.md

```diff
@@ -0,0 +1,513 @@
+  name: namespace-owner
+rules:
+  - apiGroups: [""]
+    resources: ["namespaces"]
+    resourceNames: ["$(NAMESPACE)"]
+    verbs: ["get", "list", "watch", "update", "patch", "delete"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: admin
+roleRef:
+  kind: ClusterRole
+  name: admin
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+- kind: User
+  name: "$(CREATOR)"
+  apiGroup: rbac.authorization.k8s.io
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: namespace-owner
+roleRef:
+  kind: Role
+  name: namespace-owner
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+- kind: User
+  name: "$(CREATOR)"
+  apiGroup: rbac.authorization.k8s.io
+```
+
+There is one issue that when the controller gets informed to create objects inside a new
+namespace from `NamespaceTemplate`, it has no idea who created the namespace.
+To be able to substitute `$(CREATOR)`, a `MutatingAdmissionWebhook` is involved in namespace
+creation request. 
+It adds an annotation `policy/namespace-creator=name` to the namespace being created.
+This annotation is immutable from further update/patch requests.
+
+#### Readiness of a namespace
+
+A newly created namespace is not ready before all objects in `NamespaceTemplate`
+are fully populated.
+To determine the readiness of a namespace, simply inspect `phase` in `status`
+subresource of the namespace. It's ready when the `phase` is `Active`.
+Alternatively, the `NamespaceTemplate` controller is registered as an _Initialization Controller_
+and uses `Initializer` admission control to hide the namespace before it's fully populated.
+
+### Security Consideration
+
+#### Opt-out is privileged
+
+A user who has the permission to create a namespace can put the opt-out annotation,
+though with proper RBAC setup, 
+the user may not have further permission in that namespace without being populated.
+The situation can be handled in a better way with a `ValidatingAdmissionWebhook` checking
+a specific RBAC binding:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: namespace-template-optout
+rules:
+  - apiGroups: ["policy"]
+    resources: ["namespacetemplates"]
+    verbs: ["optout"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: namespace-template-optout
+roleRef:
+  kind: ClusterRole
+  name: namespace-template-optout
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+- kind: User
+  name: "permitted-user"
+  apiGroup: rbac.authorization.k8s.io
+```
+
+The webhook will fail the request if the opt-out annotation is present without the RBAC binding for the requestor.
+
+#### Race Condition
+When a namespace is created, there's a gap before all objects in the namespace
+templates are populated.
+During the gap, the users may have the access to create objects into namespaces.
+
+To avoid the situation, a few mechanisms can be used:
+
+##### Creating namespace through service
```

Comment:
We have a similar concept in openshift via a feature called projectrequest

Reply:
After talking with jordan I'm almost convinced we can kill project request off in favor of this, with the security caveats.